Fix #1642: Missing c functions

rwhaling opened the pull request.

ekrich commented on posixlib/src/main/scala/scala/scalanative/posix/netdb.scala:
```diff
@@ -50,6 +50,8 @@ object netdb {
 
   @name("scalanative_AI_CANONNAME")
```
Just a comment since this C code already exists. I think it is much more pleasing to use all lowercase even if the C macro is all uppercase. I think we should search through on a another PR to normalize these. @LeeTibbert, do you concur?

LeeTibbert replied:
@ekrich  Yes, I agree,  The existing line (& siblings) should not be changed for this PR and
that either you or I should create an Issue about the casing of `@name(""scalanative_foo`).
I just checked posixlib and posix/sys and the usage appears to be systematically random between files.

In fact, just the other day I created a PR adding support for posix/sys/resource. I used scananative_FOO
because the models I happened to pick out of the directory all used that form and I wanted to be
consistent with them.  I think the scalanative_foo form has greater consistency with the larger scope of the project.

rwhaling replied:
I think that would look nicer also, but agree that it doesn't belong in this PR

ekrich commented on posixlib/src/main/scala/scala/scalanative/posix/sys/wait.scala:
```diff
@@ -0,0 +1,10 @@
+
+@extern
+object wait {
+  @name("wait")
```
I am really not understanding this exactly. I am seeing the following:
```
def wait(status: Ptr[CInt]: pid_t = extern
```
Reference: http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_wait.h.html

Note: `pid_t` is in sys/types.

LeeTibbert replied:
I think this is an attempt to solve the fact that the object & the method within it
have the same name.  Privately, I call this the "errno.errno" problem because I have
tripped over it so many times. My karma probably did not improve from all the bad
words I used on errno.errno.

There are a number of bad solutions.     I think the least bad is to do away with
the `@name`, use the expected `wait` as the method name and, importantly,
leave a one line comment above the method something like
`Use as wait.wait`.  I damn myself with my own writing....

rwhaling replied:
@LeeTibbert @ekrich it's actually worse than that - `wait()` collides with a method on java.lang.Object:
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html
So it can't be overridden, but I wanted to make the C wait() available anyway.

rwhaling replied:
@ekrich you are absolutely right that I totally bungled the signature of wait() lol.

LeeTibbert replied:
@rwhaling   I concur that making the C wait available is a worthwhile goal.

Aren' the method signatures different?  the C wait takes a pointer to a pid_t as
an argument and returns a pid_t.  I followed your URL (thank you) and the 
Java method(s) return void. 

As always, if I am out of order or out to lunch here, let me know.

ekrich commented on posixlib/src/main/scala/scala/scalanative/posix/sys/wait.scala:
```diff
@@ -0,0 +1,10 @@
+object wait {
+  @name("wait")
+  def _wait(): Int                                           = extern
+  def waitpid(pid: Int, status: Ptr[Int], options: Int): Int = extern
```
`waitpid` should use the CInt type.

ekrich commented on posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:
```diff
@@ -17,6 +17,7 @@ object unistd {
   def execve(path: CString, argv: Ptr[CString], envp: Ptr[CString]): CInt =
     extern
+  def execvp(path: CString, args: Ptr[CString]): CInt             = extern
   def fork(): CInt                                                = extern
```
`fork()` should return `pid_t` for readability.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/unistd.scala:
```diff
@@ -17,6 +17,7 @@ object unistd {
   def dup2(fildes: CInt, fildesnew: CInt): CInt   = extern
   def execve(path: CString, argv: Ptr[CString], envp: Ptr[CString]): CInt =
     extern
+  def execvp(path: CString, args: Ptr[CString]): CInt             = extern
```
I have an outstanding PR #1622 "WIP Fix #1621: supply complete POSIX unistd definitions" which
addresses some of these issues. We may need to manage merge conflicts.  No problem. 

I noticed that execv() seems to be missing in this edit.  The GNU execvpe() is an open question. Do we
be useful & non-Posix or strictly Posix?  

In checking the Open Group 2018 specification for this comment, I discovered that unistd.h should
declare `fexecve(int, char *const [], char *const []);`.

rwhaling replied:
@LeeTibbert the feedback I got from Denys is that only standardized POSIX functions belong in posixlib.  

I'll add `execv()` in this PR - less eager to implement `fexecve` since I've never used it, and wouldn't really know how to test it.

LeeTibbert replied:
re: execvpe Understood and agreed.   A GNUism.

re: fexecve.  I believe that unit tests for straight "passthre" methods are encouraged but
not strictly required.  I think the bar is a bit higher for those where the scala "glue" layer
does something, like handling impedance between SN and C structures.

I could not find an existing UnistdSuite.scala, so I think the lack of a fexecve() (or others
from this PR) is fine.  One voice, others may differ.

LeeTibbert replied:
OK by me if fexecve() is delayed until my PR.  May motivate me.
I was just wearing my PR review hat and pointing out inconsistencies.
Can't fix the whole world at once.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/sys/wait.scala:
```diff
@@ -0,0 +1,10 @@
+package scala.scalanative.posix.sys
```
I did not notice an associated edit to `docs/lib/posixlib.rst` to announce that `wait` is now supported.

My recent PR #1646 added support for resource.h so I had to edit posixlib.rst.  I think/hope I got it right.
Those edits might be useful to you.  The key is in knowing that there are  __two__ places to edit in
the file.     

If it would be useful to you, I could do the wait.h edits, do a sphinx-build to ensure they build (skipping over
known errors), and ship it to you.  I am trying to ease the PR flow, if possible.

rwhaling replied:
I see, I don't think I knew that document existed, I'll update this with these functions.

rwhaling replied:
it looks to me like sys/wait.h is already listed in there for some reason?

LeeTibbert replied:
I just checked the "latest" (0.4.0-M2) ReadTheDocs web output and posix sys/wait.h
is listed as "N/A" (Not Available? I'll create an separate issue ). 

As this PR implements sys/wait.h, I would expect the right column to
contain a valid link, such as the implemented sys/types.h a few lines above.
(By being concrete I mean to be descriptive, not to give offense.)

Ar my tired eyes missing something?

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/netdb.scala:
```diff
@@ -50,6 +50,8 @@ object netdb {
   @name("scalanative_AI_CANONNAME")
   def AI_CANONNAME: CInt = extern
+
+  def gai_strerror(code: CInt): CString = extern
```
If the gai_strerror() declaration moved up to just before
```
 @name("scalanative_getaddrinfo")
  def getaddrinfo(name: CString,
```
then the methods & "constants" would then each be grouped together
rather than interspersed.  The methods would also be in alphabetical order
and easier to find.

ekrich replied:
Agreed that is a good idea.

rwhaling commented:
@LeeTibbert - just getting around to reviewing your feedback, so first of all: thank you, these are great notes.  I do see that you have another PR outstanding with significantly broader scope - this PR was targeted at exactly the functions I needed to support the material included in my book, so I want to get this merged in the next few weeks.

Looking over your PR #1622, I don't actually see too much overlap, so I hope this isn't too disruptive.

rwhaling commented:
@ekrich @LeeTibbert - thank you both for the great notes.  I've pushed up the following changes:
- moved `gai_strerror` as requested
- fixed the signatures of `fork`, `wait`, and `wait_pid` to use `CInt` and `pid_t` as appropriate
- added `execv`
Y'all OK to merge?

LeeTibbert commented:
@rwhaling, @ekrich 

To my thinking the name of the ?wait() method should be settled, especially if it is
going to appear in the/your book.  I left a comment inline suggesting that with the
now current signature, the wait() name should work (untested).

The doc issue is not as critical, but worthwhile.

Since I have sphinx-build setup, I can do a follow-up PR once this is one is merged. 
Would that help move things along and remove barriers?

rwhaling commented:
@ekrich thanks for the continued eye for detail - @LeeTibbert, good call, you are right that wait() no longer conflicts with Object, and I did find the spot in `posixlibs.rst` to edit, so those are both in now.

LeeTibbert commented on docs/lib/posixlib.rst:
```diff
@@ -78,7 +78,7 @@ C Header          Scala Native Module
 `sys/un.h`_       N/A
 `sys/utsname.h`_  N/A
-`sys/wait.h`_     N/A
+`sys/wait.h`_     scala.scalanative.posix.sys.wait_
```
@rwhaling  Thank you for the doc edit and your continuing energy.
Nothing is _ever_ simple

By visual execution only, I think the new link gives information but
will fail if anybody clicks on it.     

I think you will find that you got one of two places that need change.
I made the same mistake two or so weeks ago.  If you check my in
process PR, you will see the _two_ places I needed to change.
Not to puff up my own work, but a concrete example that I 
built.

rwhaling replied:
Got it!  Thank you for your patient explanation, will do

ekrich replied:
@rwhaling It looks like you have addressed everything except the documentation change that is missing the link to the source code. @LeeTibbert Anything else?

LeeTibbert replied:
I think the missing/broken link is the only thing needed.  The rest of the PR
looks fine to my eyes.

LeeTibbert replied:
I think the missing/broken link is the only thing needed.  The rest of the PR
looks fine to my eyes.

rwhaling replied:
Thanks for taking a look at this again y'all, it totally slipped off my docket - I'll get it finished this weekend.

rwhaling commented on docs/lib/posixlib.rst:
```diff
@@ -202,6 +202,7 @@ C Header          Scala Native Module
 .. _scala.scalanative.posix.sys.time: https://github.com/scala-native/scala-native/blob/master/posixlib/src/main/scala/scala/scalanative/posix/sys/time.scala
 .. _scala.scalanative.posix.sys.types: https://github.com/scala-native/scala-native/blob/master/posixlib/src/main/scala/scala/scalanative/posix/sys/types.scala
 .. _scala.scalanative.posix.sys.uio: https://github.com/scala-native/scala-native/blob/master/posixlib/src/main/scala/scala/scalanative/posix/sys/uio.scala
+.. _scala.scalanative.posix.sys.wait: https://github.com/scala-native/scala-native/blob/master/posixlib/src/main/scala/scala/scalanative/posix/sys/wait.scala
```
@LeeTibbert this is the extra line I need right?

LeeTibbert replied:
I think all is good here.  I pulled javalib.rst. sphinx-build runs without an error related to
that file. When I use a browser (firefox) to view the generated html page, the link
is active. Clicking on it causes it to be followed.   Because it is a link to new
code, the target of the link does not yet exist, but that is an "expected" error until
this PR is merged.

LeeTibbert commented:
@rwhaling   LGTM  As a SN user, I thank you for your time & effort creating & polishing this PR.

rwhaling commented:
Thanks again @LeeTibbert! @densh - this looks good to merge now.

ekrich reviewed (approved):
It looks like this has been reviewed by myself and @LeeTibbert if it is still mergable.

LeeTibbert commented:
@rwhaling  

I believe that the issues this PR addresses have been fixed over the past two years.
Most recently PR #2969  makes `wait()` available and gives examples of how to both use and
not use (`waitpid(-1, Ptr[Int], 0)`) that method.  You and we struggled with that issue in
this PR and some of those discussions stuck with me.  That & needing to get javalib
Process.scala working reliably.

Progress happens, sometimes slowly.

I hope that you are still doing Scala work.  It is on my ToDo list to read your book and
find out how Real People do it.
